helper/random/string_generator.go

```diff
@@ -119,7 +121,10 @@ func (g *StringGenerator) generate(rng io.Reader) (str string, err error) {
 	// If performance improvements need to be made, this can be changed to read a batch of
 	// potential strings at once rather than one at a time. This will significantly
 	// improve performance, but at the cost of added complexity.
-	candidate, err := randomRunes(rng, g.charset, g.Length)
+	g.charsetLock.RLock()
+	charset := g.charset
+	g.charsetLock.RUnlock()
```

Comment:
Can you explain why locking is needed for setting a variable like this?

Reply:
The lock isn't necessary for the write to the local `charset` variable, but the read of the `g.charset` value. 

<details>

<summary> In <code>validateConfig</code>:  </summary>

```go
if len(g.charset) == 0 {
    g.charset = getChars(g.Rules)
}
```

so if there's parallel access to both `validateConfig` and `generate`, we will race reading this variable. Since this assignment isn't an atomic write (its a `[]rune` slice, which is a pointer type), you might read a bad pointer (like, half of the pointer's value could be copied over when you read, so you'd point to a garbage memory location) and thus fail.  

</details>

In particular, I think when you first initialize the group and set it up to do a password rotation, if you have a lot of in-flight requests, you'll see multiple try and write this charset via validating the config I believe, and some which will be past it trying to do this `generate` read already. 

Or maybe it is parallel modification + use of the policy...